### PR TITLE
Create unit test metrics with unit tests

### PR DESCRIPTION
**Pull request by gatli**
This PR introduces creating unit test metrics alongside the unit test. It also cleans up the ModelCI client according to our object oriented principles (sorry, should probably have made another PR 😢 ) which Phil documented earlier.

A part of this was creating a more "Pythonic" way of defining Metrics for a test.

I've introduced the class `AvailableEvalFunctions` as a collection object for `EvalFunctions`. This collection object acts as an entry to discover available standard (and in the future custom) evaluation functions through the SDK. 

Standard Scale provided evaluation functions are created as an attribute on this class and should be documented in the SDK allowing for nice IDE documentation. This allows us to define evaluation conditions (newly introduced term, plz challenge if you don't like it) for unit tests in a pythonic manner via statements like `iou() > 0.5` which then translates to the data transfer object `EvalFunctionCondition(id="ef_...", threshold_comparison=ThresholdComparison.greater_than, threshold=0.5)` which hides away the details of the REST API.

In the future this also creates a simple interface to add more parameters to evaluation functions via something like `iou(classes=["pedestrian", "cyclist"]) > 0.5`.

The new way of adding metrics to unit tests is via the `create_unit_test` method (notice the `evaluation_metrics`)
```
e = client.modelci.eval_functions
client.modelci.create_unit_test(
        name=test_name,
        slice_id=slc.slice_id,
        evaluation_metrics=[e.iou() > 0.5],
    )
```
or via `UnitTest.add_metric`
```
e = client.modelci.eval_functions
iou = e.iou
unit_test: UnitTest = ...
unit_test.add_metric(iou() > 0.5)
```

Some of the dataclasses have been translated into `pydantic.BaseModel` classes that allow us to parse the REST structures into Pythonic structures that take care of validation as well. This avoids having to create specialized `format_<class>` structures for nested dictionaries and allows us a clear place for data validation over the wire.

### CODE REVIEW

File: nucleus/modelci/__init__.py
Comment (gatli) [inline nucleus/modelci/__init__.py:96]: Moved to `UnitTest`

File: nucleus/modelci/__init__.py
Comment (gatli) [inline nucleus/modelci/__init__.py:145]: Move to `UnitTest`

File: nucleus/modelci/__init__.py
Comment (gatli) [inline nucleus/modelci/__init__.py:184]: Move to `UnitTest`

File: nucleus/modelci/__init__.py
Comment (gatli) [inline nucleus/modelci/__init__.py:289]: Moved to `UnitTest`

File: nucleus/modelci/eval_function.py
Comment (gatli) [inline nucleus/modelci/eval_function.py:1]: Moved to `nucleus/modelci/data_transfer_objects/eval_function.py`

File: nucleus/modelci/unit_test.py
Comment (gatli) [inline nucleus/modelci/unit_test.py:130]: Moved from ModelCI `client`

File: nucleus/modelci/unit_test.py
Comment (gatli) [inline nucleus/modelci/unit_test.py:144]: Moved from ModelCI `client`

File: nucleus/modelci/__init__.py
Comment (phil-scale) [inline nucleus/modelci/__init__.py:61]: nit: add tab to this - I think it needs an extra tab space to show up as code in the docs

File: tests/helpers.py
Comment (phil-scale) [inline tests/helpers.py:18]: Yeah in fact we don't need this anymore! Now that `list_eval_functions` is good, you can just use the first eval function from there.

File: nucleus/modelci/eval_functions/available_eval_functions.py
Comment (phil-scale) [inline nucleus/modelci/eval_functions/available_eval_functions.py:28]: Why not just use `BaseEvalFunction`?

File: nucleus/modelci/eval_functions/available_eval_functions.py
Comment (phil-scale) [inline nucleus/modelci/eval_functions/available_eval_functions.py:38]: I get the idea here, but I think it's a bit confusing and unscalable to have a separate class that acts as a collection of all the eval functions. I'd prefer to keep a native list for consistency since all of our other "list" methods return native lists.

File: nucleus/modelci/data_transfer_objects/eval_function.py
Comment (phil-scale) [inline nucleus/modelci/data_transfer_objects/eval_function.py:15]: We initially went with eval function name but this quickly becomes a problem because of the concept of public eval functions. For each user, we enforce uniqueness of their evaluation function names, but we do not enforce global uniqueness of names, which means that public eval functions created by other users may have conflicting names.

We could address that by also enforcing uniqueness of names among a user's eval functions and public eval functions - I think that should be the way to go.

**Review by phil-scale — COMMENTED** [review]
We're writing docstrings in a way that can look nice on the autogenerated documentation based on our comments. You can preview docs by running `poetry run sphinx-autobuild docs docs/_build/html --watch nucleus` and you can see your own docs at http://localhost:8000/api/nucleus/modelci/index.html

I think the eval functions interface is a bit confusing here because there are so many different variations of the concept of an eval function with `EvalFunctionEntry`, `BaseEvalFunction`, `AvailableEvalFunctions`, `GetEvalFunctions`, `EvalFunctionCondition`, and the union `EvalFunction`. I think we should remove a few and combine a few others if possible.

Finally, I happened to do a lot of the same refactoring here: https://github.com/scaleapi/nucleus-python-client/pull/158 but without any new functionality. I wonder if we can either merge that one into this one or merge that one into `model_ci_release` first and rebase this on top of that.

File: nucleus/modelci/eval_functions/available_eval_functions.py
Comment (phil-scale) [reply nucleus/modelci/eval_functions/available_eval_functions.py:38]: I think in this case, the functionality to map by name is sort of optional for users, and we can provide optional functionality in the form of util functions.

**Comment by phil-scale** [comment]
Also, wdyt of introducing faux immutability to pydantic classes via https://pydantic-docs.helpmanual.io/usage/models/#faux-immutability

File: nucleus/modelci/eval_functions/available_eval_functions.py
Comment (gatli) [reply nucleus/modelci/eval_functions/available_eval_functions.py:28]: MyPy wasn't happy with the base class for some reason 🤷‍♂️

File: nucleus/modelci/eval_functions/available_eval_functions.py
Comment (gatli) [reply nucleus/modelci/eval_functions/available_eval_functions.py:38]: A list or a dictionary doesn't play well at all with an IDE or Jupyter Notebook. You can't return a named entity without having an aggregator class. 

In this instance the `eval_functions` property behaves more like a namespaced module than a list ... and I think that's good and Pythonic.

When we add new evaluation functions we should add them as named objects here as well to make sure that the Jupyter notebook/IDE knows what available eval functions there are with normal autocomplete functionality. The abstraction `AvailableEvalFunctions` should barely be seen by the user, they only interact with its attributes through the modelci class.

I actually would like to change our other list methods into returning a query object as well, similar to this one. That allows for a lot more flexibility in interacting with the backend in a fast manner. 

I think this interface is very user-friendly and easier to write than more REST like APIs so I suggest we try this out and if we get bad feedback we can roll back to a simpler model.

File: nucleus/modelci/data_transfer_objects/eval_function.py
Comment (gatli) [reply nucleus/modelci/data_transfer_objects/eval_function.py:15]: Yeah, the name problem seems to keep popping up. We need to think about how we can make it as nice as we can for the user. These IDs are simply non-readable so if you're stepping through your code you have no idea what this eval function is. I don't know what the best way to go about this but we'll figure it out 🙂

**Comment by gatli** [comment]
> I think the eval functions interface is a bit confusing here because there are so many different variations of the concept of an eval function with `EvalFunctionEntry`, `BaseEvalFunction`, `AvailableEvalFunctions`, `GetEvalFunctions`, `EvalFunctionCondition`, and the union `EvalFunction`. I think we should remove a few and combine a few others if possible.

This is how you build abstraction hierarchies with  `pydantic`, this reflects how many different payload types we have communicating with the backend. Ideally we don't expose the `data_transfer_objects` to the user and rather have Model objects that they interact with. But I agree the clarity could be better. I think we just need to take another stab at the naming and granularity of these. Maybe they'll also get better when people are familiar with `pydantic`.

> Also, wdyt of introducing faux immutability to pydantic classes via https://pydantic-docs.helpmanual.io/usage/models/#faux-immutability

Very much in favour of that 🙂 


> Finally, I happened to do a lot of the same refactoring here: #158 but without any new functionality. I wonder if we can either merge that one into this one or merge that one into `model_ci_release` first and rebase this on top of that.

I merged it into `model_ci_release` and will rebase on top 🙌

**Comment by gatli** [comment]
@phil-scale I have yet to go through the Sphinx docs and merge with your changes from the releaase branch. I'll finish that tomorrow.

Just to reiterate:
I really would like to try out the `AvailableEvalFunctions` approach as I think it is pretty neat. With bigger deployments I think query objects are also the future for other "list-like" functions as currently it is pretty hard to query for certain attributes of entities.

File: nucleus/modelci/unit_test.py
Comment (phil-scale) [inline nucleus/modelci/unit_test.py:22]: I actually got rid of this class because the user never uses this, and it felt a little silly to just have a class of this form for no real purpose. I'm wondering if we can just get rid of the `info()` method because it's not super useful, and users will never call it either.

File: nucleus/modelci/unit_test.py
Comment (phil-scale) [reply nucleus/modelci/unit_test.py:128]: Yeah I think we can delete this method. It's currently unused, and users will never call it either.

File: tests/modelci/test_eval_function.py
Comment (phil-scale) [inline tests/modelci/test_eval_function.py:3]: Any way we could turn this into an assertion?

File: tests/modelci/test_unit_test.py
Comment (phil-scale) [inline tests/modelci/test_unit_test.py:20]: nit: can we also rename the variable to `criteria`?

**Review by phil-scale — COMMENTED** [review]
Overall very nice, adding a few more comments here!

File: nucleus/modelci/__init__.py
Comment (phil-scale) [inline nucleus/modelci/__init__.py:66]: As suggested by @pfmark, maybe we should force the list of `evaluation_criteria` to be nonempty

File: nucleus/modelci/eval_functions/available_eval_functions.py
Comment (pfmark) [reply nucleus/modelci/eval_functions/available_eval_functions.py:92]: I think it's a bit annoying that we need to add info about new eval functions at many different places (new class, add to `EvalFunctions`, add to `AvailableEvalFunctions`). But I guess whenever we try not to have the eval functions as member of `AvailableEvalFunctions`, IDEs won't like it, right?

File: nucleus/modelci/data_transfer_objects/eval_function.py
Comment (pfmark) [inline nucleus/modelci/data_transfer_objects/eval_function.py:8]: Shouldn't the object actually be called `EvaluationCriterion`? Then we'll also pass a `List[EvaluationCriterion]` to a unit test upon creation. It's also a `List[int]` and not a `List[intS]` ;)

File: nucleus/modelci/eval_functions/base_eval_function.py
Comment (pfmark) [inline nucleus/modelci/eval_functions/base_eval_function.py:23]: Would it also make sense to have annother abstract method which returns the eval function id `ef_...`? I think this would be a nicer interface for the user, rather than to got through the client and list them first, as we have knowledge about the specific one here (or rather in the child class)

**Review by pfmark — CHANGES_REQUESTED** [review]
Few small comments. I really like the new interface. Let's user set up tests quickly and is not error prone anymore. It's a bit annoying that we need to add eval functions at that many places, but I think we can live with that for now and rather have a neat UX. 

Let's add an assertion that we need at least one eval function in order to create a unit test. 

Will also need to adapt the docstrings then, but we can do this in another PR once this is merged. I started with that already for other existing stuff.

File: tests/modelci/test_unit_test.py
Comment (pfmark) [reply tests/modelci/test_unit_test.py:21]: As long as it's a function it should be `get_...`, if it's `criteria` only it should become a property.

**Review by pfmark — APPROVED** [review]
lgtm, if checks pass ;)

File: tests/modelci/test_eval_function.py
Comment (gatli) [reply tests/modelci/test_eval_function.py]: Changed it to check for right instance type.

File: nucleus/modelci/unit_test.py
Comment (gatli) [reply nucleus/modelci/unit_test.py:22]: Removing!

File: nucleus/modelci/__init__.py
Comment (gatli) [reply nucleus/modelci/__init__.py:68]: I've added an empty check.

**Comment by gatli** [comment]
Currently blocked by backend deployment and bugfix https://github.com/scaleapi/scaleapi/pull/33371

File: nucleus/modelci/unit_test.py
Comment (phil-scale) [reply nucleus/modelci/unit_test.py:122]: I don't think this method is necessary or belongs in `unit_test.py` now.

**Review by sasha-scale — APPROVED** [review]
Let's make sure documentation is up to date with this change!

**Comment by pfmark** [comment]
> Let's make sure documentation is up to date with this change!

Will follow in a second PR once this is merged!

**Pull request merged**
